Import export

jolavb opened the pull request:
Add KLM Import to map for trails and hydrants. Here's how it works:

In Import/Export I added a new ImportExport component and an Import Button and Input field to that container. When the input changes, it will update the SelectedFile State. When Import is clicked, the FileReader reads the SelectedFile (hopefully it's a KLM otherwise you're screwed) and passes the string into KLMReadFeatures, turning it into an OL object. From there, if mode is trails it will run ProcessTrails otherwise ProcessHydrants runs. Both are responsible for parsing the OL object and preparing it to be stored in state. Once finished parsing it passes back to the container to be stored in state. Another mode check is done there to figure out which state property to update.


Also I changed the scenery to something more beaver creeky...in doing so I realized that we should probs add a change base layer feature since a big ass cloud is covering half the mountain.

caacree commented on package.json:
```diff
@@ -3,6 +3,7 @@
   "version": "0.1.0",
   "private": true,
   "dependencies": {
+    "@mapbox/togeojson": "^0.16.0",
```
Can remove this now

jolavb replied:
Gone

caacree commented on src/components/Container.js:
```diff
@@ -131,58 +120,72 @@ class Container extends React.Component{
-    const {trails, createType, selectedTrail, hydrants} = this.state;
+    const {trails, mode, selectedTrail, hydrants} = this.state;
+
+    console.log(this.state.trails.toJS())
```
Leftover console statement

jolavb replied:
Gone

caacree commented on src/components/Container.js:
```diff
@@ -131,58 +120,72 @@ class Container extends React.Component{
-    }else {
-      this.setState({createType: type || null, selectedTrail: null});
+  importKMLClicked = (klmData) => {
+    console.log(klmData)
```
More console log

jolavb replied:
Gone

caacree commented on src/components/Container.js:
```diff
@@ -131,58 +120,72 @@ class Container extends React.Component{
-      this.setState({createType: type || null, selectedTrail: null});
+  importKMLClicked = (klmData) => {
+    console.log(klmData)
+    if( this.state.mode === 'trails'){
```
Is this necessary to distinguish between trail and hydrant KML's? 

If trails are always series of coordinates and hydrants are always points, why can't the upload function just distinguish which is which and put them in state without the user needing to be in a certain menu mode? That could allow imports of mixed trails and hydrants in the future as well without any changes.

caacree commented on src/components/Container.js:
```diff
@@ -131,58 +120,72 @@ class Container extends React.Component{
         />
+
+        <ImportExport
+        mode = {mode}
```
Nitpick: indent the props. Did you get your linter setup? It should catch this.

jolavb replied:
Indented. I have the linter setup but it doesn't like this:

```js
  renderTrailsDebounce = _.debounce(this.renderTrails, 50);
```
it throws a parsing error at the "=" and gives up linting.

caacree commented on src/components/ImportExport.js:
```diff
@@ -0,0 +1,91 @@
+import React from 'react';
+import { Button } from 'material-ui';
+import tj from '@mapbox/togeojson';
+import fs from 'fs';
```
Lots of unused imports, xmldom, tj, fs

jolavb replied:
Gooooone

caacree commented on src/components/ImportExport.js:
```diff
@@ -0,0 +1,91 @@
+    const reader = new FileReader();
+
+    reader.onload = function(event){
+      const kml = new KML().readFeatures(event.target.result)
```
Glad this ended up working cleanly!

jolavb replied:
YES. Same.

caacree commented on src/components/ImportExport.js:
```diff
@@ -0,0 +1,91 @@
+    }
+
+
+    function processTrails(klm){
```
kml, not klm. I almost wouldn't mind if we were inconsistent and wrong everywhere, but should at least be the same, haha

jolavb replied:
![alt text](https://apps.static-afkl.com/travel/us_en/images/klmlogo_new_tcm518-611911.jpg "klm not kml")

jolavb replied:
But really though, I changed it to KML

caacree commented on src/components/ImportExport.js:
```diff
@@ -0,0 +1,91 @@
+          name: feature.get('description'),
+          id: index,
+          coords: _.chunk(coords,3),
+          hydrants: null,
```
Doesn't need a hydrants attribute.

jolavb replied:
Goooooone.

caacree commented on src/components/ImportExport.js:
```diff
@@ -0,0 +1,91 @@
+
+    return (
+        <div style={style}>
+          <input onChange={this.changeFile} type='file' id='layerImport' />
```
In general, I wouldn't add id's or classes that aren't going to be used.

jolavb replied:
Goooone.

caacree commented on src/components/OpenLayersMap.js:
```diff
@@ -168,7 +179,7 @@ class OpenLayersMap extends React.Component{
     // Orientation
     var projection = Projection.get('EPSG:3857');
-    var killingtonCoords = [-72.803584,43.619210];
+    var killingtonCoords = [-106.553668, 39.612616];
```
Put this back to killington coordinates? Or we can just make Beaver Creek the default location too and change this name, doesn't matter since there aren't any default trails or hydrants anymore.

jolavb replied:
I figured it doesn't really matter since the default shouldn't be any mountain, but should be the central US or something. Changed it for a change of scenery and..see last sentence of PR message.

jolavb replied:
We should change that variable to centerCoords or something though. I'll do that in my next commit.

caacree commented on src/components/OpenLayersMap.js:
```diff
@@ -185,6 +196,8 @@ class OpenLayersMap extends React.Component{
     });
 
+
+
```
Any reason for the newlines?

jolavb replied:
Slipped and banged my head on the spacebar

jolavb replied:
Goooone.

caacree commented:
Gave it an initial pass, check it out and I'll take another look later on; also check out whatever the conflict is in Container.

caacree commented:
You can validate that a kml file is passed in; just check the extension at the end of the filename. If we were freaky about security, people COULD upload different formats and manually override the extension to try and mess with us, but for checking that people didn't accidentally load the wrong thing that's going to break everything and make them need to refresh, that should be enough.

caacree commented:
Just a thought on the git thing; are you sure you pulled the latest master to your local after my last PR and then you merged that master into your branch? I'm thinking you might be trying to merge the old version of master before mine was put in, so that's why it's making you repeat the work and still shows a conflict.

jolavb commented:
Alright, here's what I did:

1. Refactored import to be independent of mode by checking the geometry type. Also returned an object from the the process functions so that we could get rid of some code in the container function. 

2. Added some basic error handling to stop everything if no file is selected and to give an error message if the file can't be processed. 

3. Changed the center coords variable.

4. Attempted to resolve conflicts with master and push that..but it appears there are still conflicts.

caacree commented on src/components/ImportExport.js:
```diff
@@ -0,0 +1,92 @@
+      }
+
+
+    function processTrails(klm){
```
still klm here instead of kml?

jolavb replied:
Changed

caacree commented:
Cool, made a couple more comments, nothing too big I think. Starting to look good!

jolavb commented on src/components/ImportExport.js:
```diff
@@ -0,0 +1,92 @@
+
+  importFile = () =>  {
+    const { selectedFiles } = this.state;
+    const { importKMLClicked, mode } = this.props;
```
Gone

jolavb commented on src/components/ImportExport.js:
```diff
@@ -0,0 +1,92 @@
+        return {
+          name: feature.get('description'),
+          id: index,
+          coords: _.chunk(coords,3),
```
I guess we don't. I removed it and used getCoordinates as you suggested.

jolavb commented on src/components/ImportExport.js:
```diff
@@ -0,0 +1,92 @@
+      reader.onload = function(event){
+        try {
+          const kml = new KML().readFeatures(event.target.result)
+          const kmlObj = kml[0].getGeometry().getType() === 'Polygon'? processTrails(kml) : processHydrants(kml)
```
Modified with your suggested code.

jolavb commented on src/components/Container.js:
```diff
@@ -131,58 +120,60 @@ class Container extends React.Component{
-      this.setState({createType: type || null, selectedTrail: null});
-    }
+  importKMLClicked = (kmlData) => {
+    this.setState(kmlData)
```
done

jolavb commented on src/components/ImportExport.js:
```diff
@@ -0,0 +1,92 @@
+        coords.pop()
+        return {
+          name: feature.get('name'),
+          id: index,
```
Instead of using date time, I used index + name to make sure that we don't get duplicates.

jolavb commented:
Done making edits and low and behold..it still works! Lemme know if there's anything else you think we should change.

caacree commented:
Cool, I think this looks good for now. I resolved the merge conflicts and am gonna merge it.
